### PR TITLE
Implemented equals and hashcode for LazilyParsedNumber to fix

### PR DESCRIPTION
**Pull request by inder123**
https://github.com/google/gson/issues/627

### CODE REVIEW

File: gson/src/main/java/com/google/gson/internal/LazilyParsedNumber.java [hunk 1]
```diff
@@ -75,4 +75,26 @@ public String toString() {
   private Object writeReplace() throws ObjectStreamException {
     return new BigDecimal(value);
   }
-}
\ No newline at end of file
+
+  @Override
+  public int hashCode() {
+    return value == null ? 0 : value.hashCode();
```
Comment (swankjesse) [inline gson/src/main/java/com/google/gson/internal/LazilyParsedNumber.java]: value is never null

File: gson/src/main/java/com/google/gson/internal/LazilyParsedNumber.java [hunk 2]
```diff
@@ -75,4 +75,26 @@ public String toString() {
...
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (this == obj) {
+      return true;
+    }
+    if (obj == null || !(obj instanceof LazilyParsedNumber)) {
```
Comment (swankjesse) [inline gson/src/main/java/com/google/gson/internal/LazilyParsedNumber.java]: the instanceof check includes a null check

File: gson/src/main/java/com/google/gson/internal/LazilyParsedNumber.java [hunk 3]
```diff
@@ -75,4 +75,26 @@ public String toString() {
...
+      return false;
+    }
+    LazilyParsedNumber other = (LazilyParsedNumber) obj;
+    return equals(value, other.value);
+  }
+
+  private static boolean equals(Object obj1, Object obj2) {
+    if (obj1 == null) return obj2 == null;
```
Comment (swankjesse) [inline gson/src/main/java/com/google/gson/internal/LazilyParsedNumber.java]: value is never null

**Comment by swankjesse** [comment]
LGTM with some nitpicks

**Comment by inder123** [comment]
Great comments! incorporated them in the pull request.

**Pull request merged**